### ISSUE TITLE
Training Module

### ISSUE THREAD

**Issue by Mino1289** [DE50-AI-Training/ai-training-backend#10]
Add the components needed for the training:
- dataset selection
- learning rate
- number of epochs
- model architecture

Create a function with all those informations (in a dict) to train (call another..)

**Comment by ZenT0x** [comment]
bonne idée
signé Eliott

**Comment by Mino1289** [comment]
We need to define the structure of the dict. 
And which data preparation have to be done.

**Comment by Mino1289** [comment]
To create such a "super-function" we do need:

- The path to a csv file containing a dataset
  - Optional: the fraction of the dataset to use as test/validation 0.8 if not specified
  - Optional: a boolean if we "auto-clean" the dataset [data_preparation.py](https://github.com/DE50-AI-Training/ai-training-backend/blob/6a6c94ad9aa331fad6154fe16d4711aba9849625/src/trainer/datasets/data_preparation.py#L25)
  - Optional: a seed for the randomization process of split train/test dataset
- A list of all the column of the dataset acting as classification classes. (For now only classification is supported)
  - TODO:
  - For image datasets, get the column there are filepath in them.
  - For image datasets, get the transformation needed (reshape, normalize ?)
- The Architecture as defined previously [architecture.py](https://github.com/DE50-AI-Training/ai-training-backend/blob/6a6c94ad9aa331fad6154fe16d4711aba9849625/src/trainer/architectures/architecture.py#L6) and [mlp_architecture.py](https://github.com/DE50-AI-Training/ai-training-backend/blob/6a6c94ad9aa331fad6154fe16d4711aba9849625/src/trainer/architectures/mlp_architecture.py#L3)
- Selection of a loss function (TODO + Optional)
- Selection of an optimizer (TODO + Optional)
- Selection of the learning rate
- Selection of a device to run the training on (TODO)
- Selection of the number of epochs / steps to run.

This is the need, and how do we structure all those data ?
Maybe separate the part of data preparation and model training ?

Tell me @Wiqiro, @nonolintello if I missed something.

**Comment by nonolintello** [comment]
un truc dans ce genre suffirait ??

def train_supervised_model(config: dict):
    csv_path = config['csv_path']
    target_column = config['target_column']
    learning_rate = config.get('learning_rate', 0.001)
    epochs = config.get('epochs', 10)
    model_class = config['model_class']
    test_fraction = config.get('test_fraction', 0.2)
    batch_size = config.get('batch_size', 32)
    seed = config.get('seed', 42)
    device = config.get('device', 'cpu')

    # juste pour fix seed
    torch.manual_seed(seed)

    # charger/prepare le dataset
    dataset = prepare_dataset(csv_path, target_column)
    train_size = int((1 - test_fraction) * len(dataset))
    val_size = len(dataset) - train_size
    train_set, val_set = random_split(dataset, [train_size, val_size])

    train_loader = DataLoader(train_set, batch_size=batch_size, shuffle=True)
    val_loader = DataLoader(val_set, batch_size=batch_size)

    # opti et loss
    model = model_class(input_dim=dataset.input_dim, output_dim=dataset.output_dim).to(device)
    criterion = nn.CrossEntropyLoss()
    optimizer = optim.Adam(model.parameters(), lr=learning_rate)

    # entrainement :)
    for epoch in range(epochs):
        model.train()
        total_loss = 0
        for X_batch, y_batch in train_loader:
            X_batch, y_batch = X_batch.to(device), y_batch.to(device)
            optimizer.zero_grad()
            outputs = model(X_batch)
            loss = criterion(outputs, y_batch)
            loss.backward()
            optimizer.step()
            total_loss += loss.item()

        print(f"[Epoch {epoch+1}] Loss: {total_loss/len(train_loader):.4f}")

    print("Training finished.")

**Comment by nonolintello** [comment]
avec genre ce cas d'usage :

config = {
    'csv_path': 'data.csv',
    'target_column': 'label',
    'model_class': SimpleMLP,
    'learning_rate': 0.001,
    'epochs': 5,
    'batch_size': 32,
    'device': 'cpu'
}

-> train_supervised_model(config)

**Comment by Mino1289** [comment]
Je pense que ça suffirai, mais il faut pouvoir avoir + de flexibilité: 
Et si l'utilisateur veut mettre un dataset d'image ? -> pouvoir choisir la colonne avec les chemins d'accès. Et aussi faire les bons appels de fonction <DataPreparation>.read_images('col_name') + éventuelles transformations (reshape, resize, etc).
